replaced wx/xml with pugixml in rXML.cpp.

mpm11011 opened the pull request:
- This mostly affects trophy files and the manipulation thereof.
- Added pugixml to 3rdparty tools.
- Modified some VS files to accomodate pugixml.
- Large merge & squash, please let me know if there are any problems with this.

vlj commented on Utilities/rXml.cpp:
```diff
@@ -18,37 +18,50 @@ rXmlNode::rXmlNode(void *ptr)
 	if (ownPtr)
 	{
-		delete reinterpret_cast<wxXmlNode*>(handle);
+		delete reinterpret_cast<pugi::xml_node*>(handle);
```
I don't like delete reinterpret_cast, is it possible to store handle with it's type and access it as void* with a member function?

mpm11011 replied:
I would have to change the header file to reflect this. Just so I understand you correctly, you want handle to bea pugi::xml_node* (or xml_document*), and use a member function to use it as a void* when appropriate?

Bigpet replied:
yeah, you can store the type. It was just a `void*` because I was too lazy/stupid to forward declare the class (because I wanted it usable in the core without linking to wx).

mpm11011 replied:
The types for handle are now specified, and all instances of reinterpret_cast have been removed. It seems as though there is no situation where handle _needs_ to be a void*, so I'll code in the accessor, but I don't think we need it.

vlj commented on Utilities/rXml.cpp:
```diff
@@ -18,37 +18,50 @@ rXmlNode::rXmlNode(void *ptr)
+	// children() returns an iterator, so we must convert the iterator to a node pointer.
+	pugi::xml_object_range<pugi::xml_node_iterator> it = reinterpret_cast<pugi::xml_node*>(handle)->children();
+	u64 it_size = std::distance(it.begin(), it.end());
+	std::unique_ptr<pugi::xml_node[]> result(new pugi::xml_node[it_size*sizeof(pugi::xml_node)]);
```
I don't really understand what's going on here... Why do you need to create a temp array here?

mpm11011 replied:
Pugixml only seems to give back the children as a set of Iterator objects, and it would not allow me to cast or reinterpret the objects as xml_nodes. So I loop over the children to put them in contiguous memory. I will try and see if returning just the "begin()" object will suffice.

mpm11011 replied:
Aha! I needed to do *it.begin() to get a pointer to a xml_node. IIRC, Iterator objects already store their members in contiguous memory, so I removed the loop.

vlj replied:
Unfortunatly std::unique_ptr<> will be destroyed after you leave the function here so accessing elements in the array will fail ...

tambry commented:
Once you're done, you should probably squash the commits properly.

tambry commented on Utilities/rXml.cpp:
```diff
@@ -1,15 +1,15 @@
 #include "Utilities/rXml.h"
 #pragma warning(push)
-#pragma message("TODO: remove wx dependency: <wx/xml/xml.h>")
+//#pragma message("TODO: remove wx dependency: <wx/xml/xml.h>")
```
Well, might as well remove that

tambry commented on Utilities/rXml.cpp:
```diff
@@ -18,37 +18,50 @@ rXmlNode::rXmlNode(void *ptr)
 	handle = ptr;
 }
 
+// These functions go unused.
```
Then remove them?

tambry commented on rpcs3/emucore.vcxproj:
```diff
@@ -673,6 +681,9 @@
   </ItemGroup>
   <PropertyGroup Label="UserMacros" />
+  <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Release|x64'">
+    <IncludePath>.\;..\;..\asmjit\src\asmjit;..\wxWidgets\include\msvc;..\wxWidgets\include;..\wxWidgets\src\zlib;..\3rdparty\ffmpeg\WindowsInclude;..\3rdparty\ffmpeg\Windows\x86_64\Include;$(VC_IncludePath);$(WindowsSDK_IncludePath);$(UniversalCRT_IncludePath);..\3rdparty\minidx12\Include;..\3rdparty\glm;..\3rdparty\GSL\include;..\3rdparty\libpng;..\3rdparty\GL;..\3rdparty\stblib;..\3rdparty\Ope ...
```
Bad. Add it to the properties sheet.

tambry commented on rpcs3/rpcs3.vcxproj:
```diff
@@ -76,6 +76,7 @@
     <OutDir>$(SolutionDir)bin\</OutDir>
     <LinkIncremental>false</LinkIncremental>
     <RunCodeAnalysis>false</RunCodeAnalysis>
+    <IncludePath>.\;..\;..\asmjit\src\asmjit;..\wxWidgets\include\msvc;..\wxWidgets\include;..\wxWidgets\src\zlib;..\3rdparty\ffmpeg\WindowsInclude;..\3rdparty\ffmpeg\Windows\x86_64\Include;$(VC_IncludePath);$(WindowsSDK_IncludePath);$(UniversalCRT_IncludePath);..\3rdparty\minidx12\Include;..\3rdparty\glm;..\3rdparty\GSL\include;..\3rdparty\libpng;..\3rdparty\GL;..\3rdparty\stblib;..\3rdparty\Ope ...
```
Same here.

Nekotekina commented:
Sorry for introducing conflicts. Added a submodule fork https://github.com/RPCS3/pugixml

mpm11011 commented:
As per @tambry 's instructions: removed the pragma message, fixed the vcxproj files, added the include directory to rpcs3_default.props instead. Working on getting a make_shared error ironed out, and then to make sure buildbots are building pugixml.cpp correctly.

paulsapps commented on Utilities/rXml.cpp:
```diff
@@ -58,7 +71,7 @@ std::shared_ptr<rXmlNode> rXmlNode::GetChildren()
 std::shared_ptr<rXmlNode> rXmlNode::GetNext()
 {
-	wxXmlNode* result = reinterpret_cast<wxXmlNode*>(handle)->GetNext();
+	pugi::xml_node* result = &(reinterpret_cast<pugi::xml_node*>(handle)->next_sibling());
```
[-Waddress-of-temporary]
  ...result = &(reinterpret_cast<pugi::xml_node*>(handle)->next_sibling());

Not sure what this is supposed to be doing but it ain't right..

mpm11011 replied:
Sorry about that. I changed that whole function, so that warning should not appear anymore. GetNext() looks like this now:
`
pugi::xml_node result = handle->next_sibling();
	if (result)
	{
		return std::make_shared<rXmlNode>(&result);
	}
	else
	{
		return std::shared_ptr<rXmlNode>(nullptr);
	}
`

paulsapps replied:
In the else block you can just return nullptr;

mpm11011 replied:
Noted and fixed (in both places where that occurs). Thanks!

mpm11011 commented:
Error is ironed out. Before I do the commit & squash, I'm gonna make sure that the buildbots build pugixml (I included 3rdparty/pugixml/src in rpcs3/CMakeLists.txt, but it looks like Appveyor is not properly building pugixml.cpp).

@Nekotekina Should I try and use the submodule fork that you created instead of the main repo, before this commit?

AniLeo commented:
This could probably conflict with master when rebasing because pugixml was added under Utilities instead of 3rdparty
It should be moved to 3rdparty though, you could probably move it to 3rdparty while rebasing

Nekotekina commented:
Wait, what? It's in 3rdparty.

AniLeo commented:
@Nekotekina My bad. I confused it with yaml-cpp.

Nekotekina commented:
@mpm11011 I added the fork as a submodule, so I suppose you should try to use it in its current state and discard your .gitmodule changes. Tell if it doesn't work for some reason.

mpm11011 commented:
@Nekotekina I changed to RPCS3/pugixml before pushing and squashing. However, according to Appveyor, it seems that my PR has become non-mergeable. 

During my squash, I ran into many merge conflicts with VKGSRender.cpp, and I hope that I didn't screw it up trying to resolve them, but I'm not sure.

Nekotekina commented:
I have no idea why you run into conflicts in Vulkan renderer. You must rebase properly, `Merge remote-tracking branch 'refs/remotes/RPCS3/master'` must not exist at all.

mpm11011 commented:
Do you mean that I should squash all the way back to, and include, all of the `Merge remote-tracking branch 'refs/remotes/RPCS3/master'` commits (hash `a49e8a45ed8a16b4808c96a4bba5c16024c63d89`)? 

As to Vulkan: I never touched Vulkan before squashing. Roughly every other step in the rebase process, git stops to tell me to handle one-line or two-line additions in VKGSRender (which could easily be auto-merged). @vlj Do you know why this might be?

tambry commented on rpcs3/rpcs3.vcxproj:
```diff
@@ -91,7 +91,7 @@
     </ProjectReference>
     <ClCompile>
-      <AdditionalIncludeDirectories>..\3rdparty\minidx9\Include;..\Vulkan\Vulkan-LoaderAndValidationLayers\include;..\Vulkan\glslang\glslang\Public;%(AdditionalIncludeDirectories)</AdditionalIncludeDirectories>
+      <AdditionalIncludeDirectories>..\3rdparty\XAudio2_7;..\Vulkan\Vulkan-LoaderAndValidationLayers\include;..\Vulkan\glslang\glslang\Public;%(AdditionalIncludeDirectories)</AdditionalIncludeDirectories>
```
Don't modify project files. Modify properties sheet instead.

mpm11011 replied:
That came from a merge somewhere - I didn't write that myself. I'll throw in a fix if need be.

tambry commented:
It looks you haven't ever rebased or properly squashed. Most of these actually look like merge commits, instead of proper rebases. Generally for one general change, it should be in one commit and there should be no merge commits or noise from other merge jabber.

mpm11011 commented:
I have been running `git rebase -i <hash>` to squash commits. What should I do to make a "proper" rebase?

tambry commented:
Well, assuming that _upstream_ is the remote for RPCS3/rpcs3, then you should simply do _git rebase upstream/master_. This might not work with this mess, though.

mpm11011 commented:
`fatal: Needed a single revision`
`invalid upstream upstream/master`

Also tried `RPCS3/rpcs3` and `https://github.com/RPCS3/rpcs3/`

tambry commented:
@mpm11011 

> Well, assuming that upstream is the remote for RPCS3/rpcs3, ...

danilaml commented:
@mpm11011 it should be `RPCS3 master` (if you are using git for windows, if not - name could be different).

AniLeo commented:
This should probably work
git remote add upstream https://github.com/rpcs3/rpcs3.git
git pull --rebase upstream master

Then you should be able to squash/edit your commits with git rebase --interactive HEAD~2

mpm11011 commented:
@AniLeo After running the git pull command you suggested, I get this: `CONFLICT (content): Merge conflict in rpcs3/Emu/RSX/VK/VKGSRender.cpp`. The rebase is now paused and waiting for me to fix this, along with appveyor.yml. I don't know why this keeps happening to the Vulkan renderer, every time I rebase.

tambry commented:
@mpm11011 Because you merged, so now it detects those changes as your changes and those changes happen to conflict with the master. You should probably using interactive mode remove such commits, that include merge commits.

Nekotekina commented on 3rdparty/pugixml:
```diff
@@ -1 +1 @@
-Subproject commit f205aaf6e1c90f1dda943edd4f82922118a85648
+Subproject commit 8b60bbdce428af065f5501fbedd3de3c426073e4
```
The commit is changed there. I think f205aaf6e1c90f1dda943edd4f82922118a85648 must be kept if it works.

mpm11011 replied:
This must have occurred when trying to switch from zeux/pugixml to RPCS3/pugixml.

Nekotekina commented on rpcs3/Emu/RSX/VK/VKGSRender.cpp:
```diff
@@ -607,6 +607,8 @@ void VKGSRender::end()
 
+	end_command_buffer_recording();
+	execute_command_buffer(false);
 
```
Incorrect changes got mixed?

mpm11011 replied:
Probably. Should I remove those lines?

tambry commented:
There are now 3 merge commits. During interactive rebase, you should remove all the merge commits and pick the only actual commit for rebase.

paulsapps commented on Utilities/rXml.h:
```diff
@@ -12,19 +14,21 @@ struct rXmlNode
+	void *AsVoidPtr();
 
-	void *handle;
+	pugi::xml_document *handle;
```
since this is deleted in the dtor it might be better as a unique_ptr

paulsapps commented on Utilities/rXml.cpp:
```diff
@@ -1,117 +1,120 @@
+void *rXmlNode::AsVoidPtr()
 {
-	handle = reinterpret_cast<void *>(new wxXmlDocument(*reinterpret_cast<wxXmlDocument*>(other.handle)));
+	return reinterpret_cast<void*>(handle);
```
static_cast should be enough

mpm11011 replied:
Fixed.

mpm11011 commented:
Travis can't find pugixml.hpp, but I include the correct directory in CMakeLists.txt. Any ideas on what might be wrong?

danilaml commented:
@mpm11011 does it work with just `<pugixml.hpp>`instead of `<3rdparty/pugixml/src/pugixml.hpp>` ?

mpm11011 commented:
I can make another push to find out, but I don't understand. If that were the problem, why did Appveyor not fail?

danilaml commented:
@mpm11011 `#include` behavior is mostly implementation-defined. Since you added `src` dir you don't need to specify full/relative path to it in include anyway.

paulsapps commented on Utilities/rXml.h:
```diff
@@ -12,19 +14,21 @@ struct rXmlNode
 	rXmlDocument();
-	rXmlDocument(const rXmlDocument& other);
-	rXmlDocument &operator=(const rXmlDocument& other);
+	//rXmlDocument(const rXmlDocument& other);
```
Why are these now removed? Shouldn't they be declared deleted else copying one might result in a double free?

mpm11011 replied:
Not sure what you mean.

paulsapps replied:
Its a copy constructor so the code will still compile if you remove them, if a type shouldn't be copied then you need to make it explicit so it can't be done accidentally e.g  rXmlDocument(const rXmlDocument& other) = delete;

mpm11011 replied:
Ah, that makes sense. Should I do the same for the &operator= constructor as well?

vlj commented:
Is it ready for merge or does it need some more work?

mpm11011 commented:
As long as @paulsapps 's last question can be answered satisfactorily, and I update the branch because it is now out-of-date, then yes it should be fine. 

However, I just have one question: is there a way to squash/rebase merge commits without inheriting the responsibility of resolving merge conflicts in files that are well outside the scope of my PR?

tambry commented:
You don't pick the merge commits. There shouldn't be any merge commits anyways. I think I gave a basic overview of the process somewhere above.

mpm11011 commented:
Instead of picking them, I'm assuming you mean during interactive rebase, do I squash them? Or delete lines that indicate merge commits?

tambry commented:
Well they shouldn't be there, so just don't pick them (simply remove them) during the interactive rebase.

mpm11011 commented:
Removed all commits that were part of a merge from the interactive rebase. I think VulkanLoaderAndValidationLayers submodule was made dirty. I will have to submit a fix.

mpm11011 commented:
Well, neither of those fixed the problem.

danilaml commented:
I think you've got wrong changes mixed in again.
You can try creating another branch from this branch, when switching to it, when `git reset --soft RPCS3 (or however upstream is called in your case) master HEAD`. It should reset that branch to current rpcs3/master while keeping your changes (you'd have to cmmit them afterwards though).

mpm11011 commented:
`fatal: cannot do soft reset with paths.`

danilaml commented:
@mpm11011 please past full command that you used.

mpm11011 commented:
`git reset --soft RPCS3 master HEAD`

tambry commented:
Well, that's not going to work.
Try:

`git reset --soft RPCS3/master HEAD`

mpm11011 commented:
No joy.
